aesd-char-driver/main.c

```diff
@@ -230,44 +292,80 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
                 kfree(free_buffer);
             }
             data->working_buffer.size = 0;
+            PDEBUG("Size of buffer after write %ld",data->circular_buffer.buff_size);
             //update start pointer in case multiple \n present
-            start_ptr = i+1;
-            retval += mem_to_malloc;
-        }
-        if(i == count - 1)
-        {
-            //incomplete packet transmitted
-            if(copy_buffer.buffer[i] != '\n')
-            {
-                mem_to_malloc = (i-start_ptr) +1;
-                //check if working buffer size is 0, if it is malloc
-                //else realloc 
-                if(allocate_memory(&data->working_buffer, mem_to_malloc)<0)
-                {
-                    retval = -ENOMEM;
-                    
-		            //Add a goto statement. We need to free COPY_BUFFER and 
-                    //UNLOCK MUTEX   
-                    goto copy_buff_free;
-                }
-                //copy data to global buffer and wait for subsequent writes to complete the packet before enqueuing
-                memcpy((data->working_buffer.buffer + data->working_buffer.size),(copy_buffer.buffer+start_ptr),mem_to_malloc);
-                data->working_buffer.size += mem_to_malloc;
-                retval += mem_to_malloc;
-            }
+            start_ptr += delim_loc+1;
         }
+        retval += mem_to_malloc;
     }
     //free local buffer
-    copy_buff_free: kfree(copy_buffer.buffer);
     release_lock: mutex_unlock(&data->lock);
+    copy_buff_free: kfree(copy_buffer.buffer);
     ret_func: return retval;
 }
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    loff_t ret_val;
+    struct aesd_dev *data;
+    if(!filp->private_data)
+    {
+        ret_val = -EINVAL;
+        goto ret_func;
+    }
+    data = filp->private_data;
+    if (mutex_lock_interruptible(&data->lock))
+    {
+        ret_val = -EINTR;      
+        //Add a goto statement to the end
+        goto ret_func;
+    }
+    ret_val = fixed_size_llseek(filp,off,whence, data->circular_buffer.buff_size);
```

Comment:
@GuruprashanthKrishnakumar how could you modify to avoid the need to lock during the call to `fixed_size_llseek`?